File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/main/command/argument/CommandArg.java
package pize.tests.voxelgame.main.command.argument;

import pize.tests.voxelgame.server.Server;
import pize.tests.voxelgame.main.command.source.CommandSource;

public abstract class CommandArg{
    
    // Должно возвращать длину строки, которую удалось разобрать, если не удалось - 0
    public abstract int parse(String remainingChars, CommandSource source, Server server);
    
    
    public static CommandArgPosition position(){
        return new CommandArgPosition();
    }
    
    public static CommandArgPlayer player(){
        return new CommandArgPlayer();
    }
    
    public static CommandArgText text(){
        return new CommandArgText();
    }
    
    public static CommandArgWord word(){
        return new CommandArgWord();
    }
    
    public static CommandArgInteger numInt(){
        return new CommandArgInteger();
    }
    
    public static CommandArgLong numLong(){
        return new CommandArgLong();
    }
    
    public static CommandArgFloat numFloat(){
        return new CommandArgFloat();
    }
    
    public static CommandArgTime time(){
        return new CommandArgTime();
    }
    
    
    public CommandArgPosition asPosition(){
        return (CommandArgPosition) this;
    }
    
    public CommandArgPlayer asPlayer(){
        return (CommandArgPlayer) this;
    }
    
    public CommandArgText asText(){
        return (CommandArgText) this;
    }
    
    public CommandArgWord asWord(){
        return (CommandArgWord) this;
    }
    
    public CommandArgInteger asNumInt(){
        return (CommandArgInteger) this;
    }
    
    public CommandArgLong asNumLong(){
        return (CommandArgLong) this;
    }
    
    public CommandArgFloat asNumFloat(){
        return (CommandArgFloat) this;
    }
    
    public CommandArgTime asTime(){
        return (CommandArgTime) this;
    }

}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/main/command/argument/CommandArg.kt
package pize.tests.voxelgame.main.command.argument

import pize.tests.voxelgame.main.command.source.CommandSource
import pize.tests.voxelgame.server.Server

abstract class CommandArg {
    // Должно возвращать длину строки, которую удалось разобрать, если не удалось - 0
    abstract fun parse(remainingChars: String, source: CommandSource, server: Server): Int
    fun asPosition(): CommandArgPosition {
        return this as CommandArgPosition
    }

    fun asPlayer(): CommandArgPlayer {
        return this as CommandArgPlayer
    }

    fun asText(): CommandArgText {
        return this as CommandArgText
    }

    fun asWord(): CommandArgWord {
        return this as CommandArgWord
    }

    fun asNumInt(): CommandArgInteger {
        return this as CommandArgInteger
    }

    fun asNumLong(): CommandArgLong {
        return this as CommandArgLong
    }

    fun asNumFloat(): CommandArgFloat {
        return this as CommandArgFloat
    }

    fun asTime(): CommandArgTime {
        return this as CommandArgTime
    }

    companion object {
        fun position(): CommandArgPosition {
            return CommandArgPosition()
        }

        fun player(): CommandArgPlayer {
            return CommandArgPlayer()
        }

        fun text(): CommandArgText {
            return CommandArgText()
        }

        fun word(): CommandArgWord {
            return CommandArgWord()
        }

        fun numInt(): CommandArgInteger {
            return CommandArgInteger()
        }

        fun numLong(): CommandArgLong {
            return CommandArgLong()
        }

        fun numFloat(): CommandArgFloat {
            return CommandArgFloat()
        }

        fun time(): CommandArgTime {
            return CommandArgTime()
        }
    }
}
